Index: serial/tools/list_ports_linux.py
===================================================================
--- serial/tools/list_ports_linux.py	(revision 494)
+++ serial/tools/list_ports_linux.py	(working copy)
@@ -110,6 +110,14 @@
     sys_dev_path = '/sys/class/tty/%s/device/interface' % (base,)
     if os.path.exists(sys_dev_path):
         return read_line(sys_dev_path)
+
+    # USB Product Information
+    sys_dev_path = '/sys/class/tty/%s/device' % (base,)
+    if os.path.exists(sys_dev_path):
+        product_name_file = os.path.dirname(os.path.realpath(sys_dev_path)) + "/product"
+        if os.path.exists(product_name_file):
+            return read_line(product_name_file)
+
     return base
 
 def hwinfo(device):
